V3 work with moving task to task

irinadrozdova opened the pull request.

irinadrozdova committed:
- updated board items routes. Added route for checking relatives for item by board
- updated task route '/api/tasks/:taskId'. Added check for broken parentId and added check for moving task to his children
- updated TaskService. Added isItMyChildren, getAllParents, getChildrenDeep, getTaskRelatives
- updated BoardItemResource. Added resource for checkRelations
- updated board item service. Added method for checkRelations
- updated board-item-board and board-item-task views
- cleaned up in task-backlog components
- updated onDrop in task-boards.component

irinadrozdova committed:
Merge branch 'v3' into v3-work-with-moving-task-to-task

Conflicts:
	public-v2/src/app/tracker/services/board-item.service.ts

irinadrozdova committed:
wip

irinadrozdova committed:
wip

irinadrozdova committed:
cleaned up

irinadrozdova committed:
Merge branch 'v3' into v3-work-with-moving-task-to-task

Conflicts:
	public-v2/src/app/tracker/tasks/task-backlog/task-backlog.component.ts
	public-v2/src/app/tracker/tasks/task-boards/task-boards.component.ts

alexkorsun commented on app/api/boardItems/index.js:
```diff
@@ -51,12 +51,18 @@ module.exports = function (app) {
+        let itemId = req.body.item && req.body.item._id ? req.body.item._id : req.body.item;
+
+        if (req.Board._id.toString() === itemId) {
+            return res.status(403).json({error: 'You can\'t added item into itself'});
```
'You can\'t add item into itself'

alexkorsun commented on app/api/task/index.js:
```diff
@@ -253,16 +253,29 @@ module.exports = function (app) {
-                        res.json(task);
+
+                        if (isChildren) {
+                            return res.status(403).json({error: 'You can\'t move task to his children'});
```
'You can\'t move task to its children'

alexkorsun commented on app/services/TaskService.js:
```diff
@@ -811,6 +805,49 @@ var TaskService = function (Task, FileService, UserService, SocketService, Histo
+                return next(err);
+            }
+
+            if (!futureParent.parentTaskId) {
```
if there is no futureParent in db? for ex. it was removed. need to check this.

alexkorsun commented on app/services/TaskService.js:
```diff
@@ -811,6 +805,49 @@ var TaskService = function (Task, FileService, UserService, SocketService, Histo
+        });
+    };
+
+    this.getChildrenDeep = (task, allChildren = [], next) => {
```
What did you write this method for? Why not to use this.deepFindByQuery?

alexkorsun commented on app/api/board/index.js:
```diff
@@ -70,5 +72,43 @@ module.exports = function (app) {
+            .then((items) => {
+                async.each(items, (boardItem, callback) => {
+                    allChildren.push(boardItem.item._id);
+                    TaskService.getChildrenDeep(boardItem.item, [], (err, children) => {
```
Why not to use this.deepFindByQuery?

alexkorsun commented on app/services/board/BoardService.js:
```diff
@@ -183,6 +183,15 @@ let BoardService = function (Board,
 
+    this.hasRelatives = (item, children) => {
+        item = item && item._id ? item._id : item;
+        children = children && children._id ? children._id : children;
```
it's array!

alexkorsun commented on public-v2/src/app/tracker/tasks/task-boards/task-boards.component.ts:
```diff
@@ -132,13 +133,67 @@ export class TaskBoardsComponent implements OnInit, OnDestroy {
+    let dropZone = dropData.params;
+    let newBoardItem = _.pick(dropData.item, ['board', 'item', 'type', '_id']);
+
+    let ifBoardToBoard = dropZone.type === 'board' && newBoardItem.type === 'board';
```
move checks which used few times to variables

irinadrozdova committed:
- fixed getting children, now use deepFindByQuery
 - fixed texts for errors
 - fixed mapping for children in hasRelatives
 - added check for 'futureParent' and removed getChildrenDeep method from TaskService
 - fixed toast.service
 - updated onDrop method in task-boards.component. Added variables for the same checks

irinadrozdova committed:
Merge branch 'v3' into v3-work-with-moving-task-to-task
